File: src/kotlin/ktor/src/main/kotlin/com/benchmark/routes/DatabaseRoutes.kt
package com.benchmark.routes

import com.benchmark.models.UserOrderStats
import com.benchmark.services.DatabaseService
import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import kotlinx.serialization.Serializable

// Both handlers used to concatenate their JSON by hand into a StringBuilder or
// a template string, with snake_case keys and an {user_id, email, order_count,
// total_amount, average_amount} shape nothing else emitted. They now go
// through the same kotlinx.serialization path as the rest of the app.
// See contracts/rest/canonical-payloads.md.

@Serializable
private data class ComplexResponse(
    val periodDays: Int,
    val totalUsers: Int,
    val data: List<UserOrderStats>,
)

@Serializable
private data class ErrorResponse(val error: String)

fun Route.databaseRoutes(dbService: DatabaseService) {
    get("/db/simple") {
        val id = call.request.queryParameters["id"]?.toIntOrNull() ?: 1
        val user = dbService.findUserById(id)
        if (user != null) {
            call.respond(user)
        } else {
            call.respond(HttpStatusCode.NotFound, ErrorResponse("User with id $id not found"))
        }
    }

    get("/db/complex") {
        val days = call.request.queryParameters["days"]?.toIntOrNull() ?: 30
        if (days <= 0 || days > 365) {
            call.respond(HttpStatusCode.BadRequest, ErrorResponse("days must be between 1 and 365"))
            return@get
        }
        val data = dbService.findComplexOrders(days)
        call.respond(ComplexResponse(periodDays = days, totalUsers = data.size, data = data))
    }
}

File: src/dart/vaden/lib/routes/database_routes.dart
import 'package:vaden/vaden.dart';
import '../models/complex_order_result.dart';
import '../services/database_service.dart';

List<Route> databaseRoutes() {
  return [
    Route.get('/db/simple', _simpleDbHandler),
    Route.get('/db/complex', _complexDbHandler),
  ];
}

Future<Response> _simpleDbHandler(Request request) async {
  final databaseService = request.context['databaseService'] as DatabaseService;

  final idParam = request.url.queryParameters['id'];
  if (idParam == null) {
    return Response.json(
      statusCode: 400,
      body: {
        'error': 'Bad Request',
        'message': 'id parameter is required',
      },
    );
  }

  final userId = int.tryParse(idParam);
  if (userId == null) {
    return Response.json(
      statusCode: 400,
      body: {
        'error': 'Bad Request',
        'message': 'id must be a number',
      },
    );
  }

  final user = await databaseService.getUser(userId);
  if (user == null) {
    return Response.json(
      statusCode: 404,
      body: {
        'error': 'Not Found',
        'message': 'User not found',
      },
    );
  }

  return Response.json(
    body: user.toJson(),
  );
}

Future<Response> _complexDbHandler(Request request) async {
  final databaseService = request.context['databaseService'] as DatabaseService;

  final daysParam = request.url.queryParameters['days'];
  final days = daysParam != null ? int.tryParse(daysParam) ?? 30 : 30;

  if (days <= 0 || days > 365) {
    return Response.json(
      statusCode: 400,
      body: {
        'error': 'Bad Request',
        'message': 'days must be between 1 and 365',
      },
    );
  }

  final users = await databaseService.getComplexUsers(days);

  final result = ComplexQueryResult(
    periodDays: days,
    totalUsers: users.length,
    data: users.map((userData) {
      return UserStats(
        userId: userData['user_id'] as int,
        userName: userData['user_name'] as String,
        totalOrders: userData['total_orders'] as int,
        totalValue: userData['total_value'] as double,
        averageValue: userData['average_value'] as double,
      );
    }).toList(),
  );

  return Response.json(
    body: result.toJson(),
  );
}
